Load the class node:

```
sinteractive --reservation=aneq505 --time=01:00:00 --partition=amilan --nodes=1 --ntasks=6 --qos=normal
```

Load your Qiime environment:
```
module purge  

module load qiime2/2024.10_amplicon
```

Let's navigate to the correct place. Use `cd` to move into your **decomp****_tutorial** folder. Use `pwd` to make sure you're in the correct place.

```
cd /scratch/alpine/$USER/decomp_tutorial
  
pwd
```

**Create directories for the different analyses so you can keep your files organized:**

```
mkdir slurm

mkdir taxonomy

mkdir tree

mkdir taxaplots

mkdir dada2

mkdir demux

mkdir core_metrics
```

qza = qiime artifact so qiime can read through it

**Create a directory for the manifest and import the manifest files (we made them for you):** 

In your study you would need to construct the manifest file yourself, 

```
mkdir manifest  

cd manifest  

cp /pl/active/courses/2025_summer/CSU_2025/q2_workshop_final/QIIME2/manifest_run2.txt .

cp /pl/active/courses/2025_summer/CSU_2025/q2_workshop_final/QIIME2/manifest_run3.txt .
```

Move out of the manifest directory:
```
cd ../
```

**Copy the demultiplex reads into your analysis directory:**

```
cp -r /pl/active/courses/2025_summer/CSU_2025/q2_workshop_final/QIIME2/reads_run2 .

cp -r /pl/active/courses/2025_summer/CSU_2025/q2_workshop_final/QIIME2/reads_run3 .
```



Let's take a look at the data we just imported, we can do this by moving into the reads folder, and typing ls
```
cd reads_run2

ls
```
these are the individual sample fastq files!

```
cd ../
```

**Import data into QIIME2:**
# run the import command to generate a qiime2 readable artifact for the reads, save the demux results in the demux folder.
```
qiime tools import \  
--type "SampleData[PairedEndSequencesWithQuality]" \  
--input-format PairedEndFastqManifestPhred33V2 \  
--input-path manifest/manifest_run2.txt \  
--output-path demux/demux_run2.qza

qiime tools import \  
--type "SampleData[PairedEndSequencesWithQuality]" \  
--input-format PairedEndFastqManifestPhred33V2 \  
--input-path manifest/manifest_run3.txt \  
--output-path demux/demux_run3.qza
```

**Check Data Quality:** 

The first important step of data analysis is to check the data quality. Now that we've gotten our data into QIIME2 it's simple to get a visualization that we can use to evaluate our sample quality.  

# generate a demux visualization file  
```
cd demux  
  
#run2  
qiime demux summarize \  
--i-data demux_run2.qza \  
--o-visualization demux_run2.qzv  
  
#run3  
qiime demux summarize \  
--i-data demux_run3.qza \  
--o-visualization demux_run3.qzv  
```



Just like above, transfer this output file to your local computer and visualize it in [view.qiime2.orgLinks to an external site.](http://view.qiime2.org/ "(opens in a new window)"). 

**Questions to Answer with this Output:**

- What is the per-sample sequencing count?
- Which sample had the fewest sequences per sample? The most?
- How many samples were in this run?
- How does the quality look?
- Does your data look the exact same as your neighbors?
- How long are the reads? 
- **Where would you trim and/or truncate these sequences to maintain the highest quality data?**